include/arch/x86/arch/machine/timer.h

```diff
@@ -61,7 +62,9 @@ static inline void setDeadline(ticks_t deadline)
         x86_wrmsr(IA32_TSC_DEADLINE_MSR, deadline);
     } else {
         /* convert deadline from tscKhz to apic khz */
-        deadline -= getCurrentTime();
+        ticks_t now = getCurrentTime(); /* ToDO: just use ksCurTime? */
```

Comment:
should this use ksCurTime here instead of getting an update time stamp from the timer driver?

Reply:
No, because this is programming a timeout in the future by a countdown timer.  So if it uses ksCurTime (which is less than getCurrentTime()), then the timeout will actually be  late because of the amount of time passed while the kernel was executing after calling getCurrentTime previously.

Reply:
setDeadline() is only called by setNextInterrupt() which does:

``time_t next_interrupt = NODE_STATE(ksCurTime) + refill_head(NODE_STATE(ksCurThread)->tcbSchedContext)->rAmount;``

and then immediately calls setDeadline(), so in practise you can as well use ksCurTime. Perhaps the API needs to be changed to make this explicit though.

Reply:
setNextInterrupt() is called very near the end of the kernel's execution whereas ksCurTime is updated upon entry to the kernel (or at preemption points).  We need to ensure that an interrupt is fired so that the latest next time the kernel is entered ksCurTime is updated to a value that isn't greater than `NODE_STATE(ksCurTime) + refill_head(NODE_STATE(ksCurThread)->tcbSchedContext)->rAmount`.  If there has been `t` time since ksCurTime was updated at the point when we are programming the timer, then an interrupt programmed to be `refill_head(NODE_STATE(ksCurThread)->tcbSchedContext)->rAmount` in the future will be late by `t` amount. `t` could be as large as the kernel's WCET, which I believe would cause the interrupt to come in 1 WCET late.

Reply:
Thanks for the clear explanation.